js-packages/@prestojs/viewmodel/src/ViewModelCache.ts

```diff
@@ -395,6 +647,101 @@ class RecordCache<ViewModelClassType extends ViewModelConstructor<any, any, any>
     }
 }
 
+/**
+ * Responsible for handling the batching of changes for the purposes of calling listeners
+ */
+const defaultListenerBatcher = {
+    /**
+     * Whether batch is in progress
+     */
+    isActive: false,
+    /**
+     * This is items that need to be notified only once per change in a batch
+     */
+    pending: new Map(),
+    /**
+     * This is items that don't need to respect batching rules. This is really for internal
+     * use but is part of the public API. This exists as to make listeners work across
+     * related fields we listen to changes on the related field cache
+     */
+    pendingNoBatch: new Map(),
+    /**
+     * This is for callbacks that only listen to all changes on a model (not specific records or fields)
+     */
+    pendingAll: new Set(),
+
+    /**
+     * Queue a call to a non-record/field specific listeners
+     */
+    callAll(listener: AllChangesListener): void {
+        if (this.isActive) {
+            this.pendingAll.add(listener);
+        } else {
+            listener();
+        }
+    },
+    /**
+     * Queue a call to a record/field specific listener
+     *
+     * If `shouldBatch` is false then there's no guarantees about the listener only being called
+     * once.
+     */
+    call<T>(
+        listener: ChangeListener<T>,
+        before?: T | null,
+        after?: T | null,
+        shouldBatch = true
+    ): void {
+        if (this.isActive) {
+            const p = shouldBatch ? this.pending : this.pendingNoBatch;
+            let listenerPending = p.get(listener);
+            if (!listenerPending) {
+                p.set(listener, [before, after]);
+            } else {
+                listenerPending[1] = after;
+            }
+        } else {
+            listener(before, after);
+        }
+    },
+    /**
+     * Start a batch. The passed function is called and any changes queued. Once function returns listeners
+     * will be dispatched. The value returned from `run` will be returned.
+     *
+     * If any error occurs in `run` then no listeners will be called.
```

Comment:
See comment later on: what is expected to happen if you try to nest batches?

(implementation doesn't seem to handle this, but even if it did then there's a question of when the notification should take place, esp if the same record is modified in both an inner and outer batch)

Reply:
the implementation handles it - batcher is just a singleton if you call it while it's already in a batch it just calls immediately:

```
if (this.isActive) {
    return run();
}
```

the notification happens at end of outer most batch

Reply:
"the notification happens at end of outer most batch" - that should be documented